src/Npgsql/ConnectorPool.cs

```diff
@@ -472,23 +484,71 @@ void ReleaseOneWaiter()
                     break;
         }
 
+        // Manual reactivation of timer happens in callback
+        void EnablePruning()
+        {
+            lock (_pruningTimer)
+            {
+                _pruningTimerEnabled = true;
+                _pruningTimer.Change(_pruningSamplingInterval, Timeout.InfiniteTimeSpan);
+            }
+        }
+
+        void DisablePruning()
+        {
+            lock (_pruningTimer)
+            {
+                _pruningTimer.Change(Timeout.Infinite, Timeout.Infinite);
+                _pruningSampleIndex = 0;
+                _pruningTimerEnabled = false;
+            }
+        }
+
         static void PruneIdleConnectors(object? state)
         {
             var pool = (ConnectorPool)state!;
-            var idle = pool._idle;
-            var now = DateTime.UtcNow;
-            var idleLifetime = pool.Settings.ConnectionIdleLifetime;
+            var samples = pool._pruningSamples;
+            int toPrune;
+            lock (pool._pruningTimer)
+            {
+                // Check if we might have been contending with DisablePruning.
+                if (!pool._pruningTimerEnabled)
+                    return;
+
+                var sampleIndex = pool._pruningSampleIndex;
+                if (sampleIndex < pool._pruningSampleSize)
+                {
+                    samples[sampleIndex] = pool.State.Idle;
+                    pool._pruningSampleIndex = sampleIndex + 1;
+                    pool._pruningTimer.Change(pool._pruningSamplingInterval, Timeout.InfiniteTimeSpan);
+                    return;
+                }
+
+                // Reset before releasing the lock and starting the prune
+                toPrune = Median(samples);
+                pool._pruningSampleIndex = 0;
+                pool._pruningTimer.Change(pool._pruningSamplingInterval, Timeout.InfiniteTimeSpan);
+            }
 
+            var idle = pool._idle;
             for (var i = 0; i < idle.Length; i++)
             {
-                if (Volatile.Read(ref pool.State.Open) <= pool._min)
+                if (Volatile.Read(ref pool.State.Open) <= pool._min || toPrune == 0)
                     return;
 
                 var connector = idle[i];
-                if (connector == null || (now - connector.ReleaseTimestamp).TotalSeconds < idleLifetime)
+                if (connector == null || Interlocked.CompareExchange(ref idle[i], null, connector) != connector)
                     continue;
-                if (Interlocked.CompareExchange(ref idle[i], null, connector) == connector)
-                    pool.CloseConnector(connector, true);
+
+                toPrune -= 1;
+                pool.CloseConnector(connector, true);
+            }
+
+            int Median(int[] values)
+            {
+                var i = (int)Math.Ceiling((double)(values.Length - 1) / 2);
```

Comment:
In which case `i` can be negative? As I see `pruningInterval` should be greater or equal to `_pruningSamplingInterval`, and there should be at least one element in the array. Right?

Reply:
True, though this is so inconsequential for perf (runs once per cycle and on the timer thread) that I'd rather have a guarded implementation here.

Reply:
Since there is no need in guard, you can replace all floating point math with integer:

```csharp
static int Divide(int value, int divisor) => 1 + (value - 1) / divisor;
```

The method can be used in two places where is `Math.Ceiling`, and after this one `Median` can be fully inlined.

Reply:
One more note - the value of i never changes, since the length of values is fixed, so we can calculate once at initialization and reuse that. I agree that all this is totally irrelevant for perf, although it might make the code a bit easier to read. Up to you, I'm fine with whatever.